# Анализ базы данных сервиса для чтения книг

В нашем распоряжении база данных крупного сервиса для чтения книг по подписке.    
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

**ЦЕЛЬ ИССЛЕДОВАНИЯ** - анализ базы данных с целью формулирования ценностного предложения для нового продукта.

**СТРУКТУРА БАЗЫ ДАННЫХ**

![](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2)

База данных состоит из 5 таблиц:

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

После подключения к базе данных через Python (код скрыт от читателя), проведем исследование таблиц и ответим на вопросы с помощью SQL-запросов.


### Исследование таблиц

Исследуем таблицы, которые есть в базе данных. Для этого c помощью SQL-запросов выгрузим таблицы в переменные, а с помощью библиотеки pandas посмотрим на число пропусков и наличие дубликатов в них (используем функцию `data_info`).

In [3]:
#создаем функцию для предобработки данных
def data_info(df):
    # выводим первые 5 строк датафрема
    display(df.head())
    # применяем метод info()
    print()
    print('Общая информация:')
    print(df.info())
    print()
    # находим число строк-дубликатов в датафреме
    print('Количество строк-дубликатов:', df.duplicated().sum())

**Таблица "books"**

In [4]:
query = '''SELECT * 
           FROM books'''

books = pd.io.sql.read_sql(sql=text(query), con = engine)
data_info(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None

Количество строк-дубликатов: 0


**Таблица "authors"**

In [5]:
query = '''SELECT * 
           FROM authors '''
           

authors = pd.io.sql.read_sql(sql=text(query), con = engine)
data_info(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None

Количество строк-дубликатов: 0


**Таблица "publishers"**

In [6]:
query = '''SELECT * 
           FROM publishers '''
           
publishers = pd.io.sql.read_sql(sql=text(query), con = engine)
data_info(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None

Количество строк-дубликатов: 0


**Таблица "ratings"**

In [7]:
query = '''SELECT * 
           FROM ratings '''
           
ratings = pd.io.sql.read_sql(sql=text(query), con = engine)
data_info(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None

Количество строк-дубликатов: 0


**Таблица "reviews"**

In [8]:
query = '''SELECT * 
           FROM reviews '''
           

reviews = pd.io.sql.read_sql(sql=text(query), con = engine)
data_info(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None

Количество строк-дубликатов: 0


Пропущенных значений в данных нет, явных дубликатов тоже. Можем переходить к решению  заданий.

### Решение заданий

#### Задание 1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [9]:
query = '''SELECT COUNT(book_id) AS quantity
           FROM books
           WHERE publication_date > '2000-01-01'; '''

pd.io.sql.read_sql(sql=text(query), con = engine)

quantity
0       819

Таблица с книгами хранит информацию о 819 книгах, опубликованных после 1 января 2000 года.

#### Задание 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [10]:
query = '''SELECT b.book_id,
                  b.title,
                  ROUND(AVG(rat.rating), 2) AS avg_rating,
                  COUNT(DISTINCT rev.review_id) AS quantity_review
           FROM books AS b
           LEFT OUTER JOIN ratings AS rat ON b.book_id = rat.book_id
           LEFT OUTER JOIN reviews AS rev ON b.book_id = rev.book_id
           GROUP BY b.book_id
           ORDER BY quantity_review DESC;
'''

pd.io.sql.read_sql(sql=text(query), con = engine)

book_id                                              title  avg_rating  \
0        948                            Twilight (Twilight  #1)        3.66   
1        963                                Water for Elephants        3.98   
2        734                                   The Glass Castle        4.21   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...        4.41   
4        695  The Curious Incident of the Dog in the Night-Time        4.08   
..       ...                                                ...         ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel        3.67   
996      808                            The Natural Way to Draw        3.00   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites        5.00   
998      221                          Essential Tales and Poems        4.00   
999      191  Disney's Beauty and the Beast (A Little Golden...        4.00   

     quantity_review  
0                  7  
1                  6  
2                  6  
3                  6  
4                  6  
..               ...  
995                0  
996                0  
997                0  
998                0  
999                0  

[1000 rows x 4 columns]

В таблице с книгами собрана информация о 1000 книг. В результате запроса получили средние рейтинги каждой книги и число обзоров, написанных на книгу.   
Книга с самым низким средним рейтингом -  "Harvesting the Heart" (идентификатор книги 303), средний рейтинг равен 1,5. Максимальный рейтинг книг составляет 5.0.    
Что касается числа обзоров, то больше всего их написано на книгу "Twilight (Twilight #1)" ("Сумерки"). Причем можно заметить, что число обзоров, на первый взгляд, со средним рейтингом не связано: встречаются книги без обзоров с максимальным рейтингом, и наоборот: у книги с максимальным числом обзоров может быть не высокий рейтинг.

#### Задание 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [11]:
query = '''
           SELECT n.publisher_id,
                  p.publisher,
                  n.number_of_books
                  
           FROM (SELECT publisher_id,
                        COUNT(book_id) AS number_of_books
                FROM books
                WHERE num_pages > 50
                GROUP BY publisher_id
                ORDER BY number_of_books DESC
                LIMIT 1) AS n
                
           JOIN publishers AS p ON n.publisher_id=p.publisher_id;
           
'''

pd.io.sql.read_sql(sql=text(query), con = engine)

publisher_id      publisher  number_of_books
0           212  Penguin Books               42

Самым активным по выпуску книг с объемом более 50 страниц стало британское издательство "Penguin Books". Число выпущеннык книг - 42.

#### Задание 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [12]:
query = '''SELECT a.author_id,
                  a.author,
                  ROUND(AVG(rating),3) AS avg_rating
                  
           FROM (--книги, у которых 50 и более оценок
                SELECT book_id
                FROM ratings
                GROUP BY book_id
                HAVING COUNT(rating_id) >= 50) AS i
                
           INNER JOIN books AS b ON i.book_id = b.book_id
           INNER JOIN authors AS a ON b.author_id = a.author_id
           INNER JOIN ratings AS r ON i.book_id = r.book_id
           
           GROUP BY a.author_id, a.author
           ORDER BY avg_rating DESC
           LIMIT 1;            
                     
'''

pd.io.sql.read_sql(sql=text(query), con = engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré       4.287

Самым рейтинговым автором можно считать тандем из британской писательницы Джоан Роулинг(J.K. Rowling) и американского иллюстратора Мэри ГрандПре (Mary GrandPré) - известными созданием книг о Гарри Поттере. Средний рейтинг авторов составил 4.287.

#### Задание 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [13]:
query = '''SELECT ROUND(AVG(i.number_of_review),0) AS avg_number_of_review

           FROM (--находим число отзывов для пользователей, которые поставили больше 50 оценок
                 SELECT COUNT(review_id) AS number_of_review
                 FROM reviews
                 WHERE username IN (--отбираем пользователей, которые поставили больше 50 оценок
                                    SELECT username
                                    FROM ratings
                                    GROUP BY username
                                    HAVING COUNT(rating_id) > 50)
                 GROUP BY username) AS i;
                            
'''

pd.io.sql.read_sql(sql=text(query), con = engine)

avg_number_of_review
0                  24.0

Пользователи, которые ставят более 50 оценок - в среднем, пишут 24 обзора.

### Выводы

* В базе данных хранится информация о 1000 книг;    
* Подавляющее большинство книг опубликованы позже 2000 года (819 штук);    
* Средние рейтинги книг варьируют от 1.5 до 5.0, самый низкий средний рейтинг у книги "Harvesting the Heart";    
* Больше всего обзоров написано на книгу "Twilight" - 7 штук, но также есть книги, на которые не написано ни одного обзора;   
* Самым активным издательством стало британское издательство "Penguin Books", выпустившее 42 книги объемом более 50 страниц; 
* Джоан Роулинг(J.K. Rowling) - самый рейтинговый автор (средняя оценка - 4.287);    
* Активные пользователи, которые ставят более 50 оценок - в среднем, пишут 24 обзора.
